console_backend/src/output.rs

```diff
@@ -50,42 +50,59 @@ impl CsvLogging {
 pub type SbpLogging = cc::SbpLogging;
 
 #[derive(Debug)]
-pub enum SbpLogger {
-    Sbp(File),
-    Json(JsonEncoder<File, CompactFormatter>),
+pub struct SbpLogger<W> {
+    logger: SbpLogging,
+    out: W,
+    path: Option<PathBuf>,
 }
 
-impl SbpLogger {
-    pub fn new_sbp<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Sbp(File::create(filepath)?))
+impl<W: Write> SbpLogger<W> {
+    pub fn new(logger: SbpLogging, out: W, path: Option<PathBuf>) -> Self {
+        Self { logger, out, path }
     }
 
-    pub fn open_sbp<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Sbp(
-            OpenOptions::new().append(true).open(filepath)?,
-        ))
-    }
+    /// Log data into respective outputs, SBP or JSON format
+    ///
+    /// # Parameters:
+    /// - `frame`: The raw frame data that a message must have
+    /// - `msg`: Optional SBP depending on validity, only required in JSON format
+    ///
+    /// # Returns: the bytes serialized and whether path exists.
+    /// Error returns should be considered interruptions which breaks the callers flow.
+    pub fn serialize(&mut self, frame: &Frame, msg: Option<&Sbp>) -> Result<u16> {
+        if msg.is_none() {
+            error!("(SBP) message cannot be parsed as SBP, serializing frame instead: {frame:?}");
+        }
+        let bytes = match &self.logger {
+            SbpLogging::SBP_JSON => msg
+                .map(to_vec)
+                .and_then(|ret| ret.ok_or_log(|_| error!("error serializing SBP to JSON"))),
+            SbpLogging::SBP => Some(frame.as_bytes().to_owned()),
+        };
 
-    pub fn new_sbp_json<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Json(JsonEncoder::new(
-            File::create(filepath)?,
-            CompactFormatter,
-        )))
-    }
+        if let Some(path) = &self.path {
```

Comment:
not sure how costly this is checking everytime we serialize.
looks like TOCTOU 👀 but it should be ok

or alternatively use some sort of file notify.

Reply:
I think using something like [notify](https://docs.rs/notify/5.0.0/notify/) or [watchexec](https://docs.rs/watchexec/latest/watchexec/) is a good idea but may require a bigger refactor than this bug necessitates. At the same time I don't think this check actually prevents all the bugs you are trying to prevent here, and it may create some new ones. You are just checking if the path exists, which could happen if someone deleted the file and then recreated it, this check would pass, but the write_all path would drop.

Another issue is that if you rename the file with `mv` the code before this change will keep writing to the renamed file, whereas this change will cause an unnecessary error imo. 

Maybe these are tradeoffs that are ok to make, I don't have the context here and the jira tickets are minimal

Reply:
yea, this would address file deletion but file moves and instantly recreated would be kinda edge case (can fix by just ensuring metadata / size of file path before writing as well - not sure how much performance that is tho) 🤔